readme.md
# Gitalk

[![NPM][npm-version-image]][npm-version-url]
[![CDNJS][cdnjs-version-image]][cdnjs-version-url]
[![jsdelivr](https://data.jsdelivr.com/v1/package/npm/gitalk/badge)](https://www.jsdelivr.com/package/npm/gitalk)
[![david-dm][david-dm-image]][david-dm-url]
[![travis][travis-image]][travis-url]
[![coveralls][coveralls-image]][coveralls-url]
[![gzip-size][gzip-size]][gzip-url]

Gitalk 是一个基于 GitHub Issue 和 Preact 开发的评论插件。

## 特性

- 使用 GitHub 登录
- 支持多语言 [en, zh-CN, zh-TW, es-ES, fr, ru, de, pl, ko, fa, ja]
- 支持个人或组织
- 无干扰模式（设置 distractionFreeMode 为 true 开启）
- 快捷键提交评论 （cmd|ctrl + enter）

[Readme](https://github.com/gitalk/gitalk/blob/master/readme.md)
[在线示例](https://gitalk.github.io)

## 安装

两种方式

- 直接引入

```html
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css" />
<script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js"></script>

<!-- or -->

<link rel="stylesheet" href="https://unpkg.com/gitalk/dist/gitalk.css" />
<script src="https://unpkg.com/gitalk/dist/gitalk.min.js"></script>
```

- npm 安装

```sh
npm i --save gitalk
```

```js
import 'gitalk/dist/gitalk.css'
import Gitalk from 'gitalk'
```

## 使用

首先，您需要选择一个公共 github 存储库（已存在或创建一个新的 github 存储库）用于存储评论，

然后需要创建 **GitHub Application**，如果没有 [点击这里申请](https://github.com/settings/applications/new)，`Authorization callback URL` 填写当前使用插件页面的域名。

最后, 您可以选择如下的其中一种方式应用到页面：

### 方式 1

添加一个容器：

```html
<div id="gitalk-container"></div>
```

用下面的 Javascript 代码来生成 gitalk 插件：

```js
var gitalk = new Gitalk({
  clientID: 'GitHub Application Client ID',
  clientSecret: 'GitHub Application Client Secret',
  repo: 'GitHub repo',
  owner: 'GitHub repo owner',
  admin: ['GitHub repo owner and collaborators, only these guys can initialize github issues'],
  id: location.pathname, // Ensure uniqueness and length less than 50
  distractionFreeMode: false, // Facebook-like distraction free mode
})

gitalk.render('gitalk-container')
```

### 方式 2：在 React 使用

使用以下代码引入 Gitalk 组件

```jsx | pure
import GitalkComponent from 'gitalk/dist/gitalk-component'
```

按以下方式在 React 中使用 Gitalk 组件

```jsx | pure
<GitalkComponent
  options={{
    clientID: '...',
    // ...
    // 设置项
  }}
/>
```

## 设置

- **clientID** `String`

  **必须**. GitHub Application Client ID.

- **clientSecret** `String`

  **必须**. GitHub Application Client Secret.

- **repo** `String`

  **必须**. GitHub repository.

- **owner** `String`

  **必须**. GitHub repository 所有者，可以是个人或者组织。

- **admin** `Array`

  **必须**. GitHub repository 的所有者和合作者 (对这个 repository 有写权限的用户)。

- **id** `String`

  Default: `location.href`.

  页面的唯一标识。长度必须小于 50。

- **number** `Number`

  Default: `-1`.

  页面的 issue ID 标识，若未定义`number`属性则会使用`id`进行定位。

- **labels** `Array`

  Default: `['Gitalk']`.

  GitHub issue 的标签。

- **title** `String`

  Default: `document.title`.

  GitHub issue 的标题。

- **body** `String`

  Default: `location.href + header.meta[description]`.

  GitHub issue 的内容。

- **language** `String`

  Default: `navigator.language || navigator.userLanguage`.

  设置语言，支持 [en, zh-CN, zh-TW, es-ES, fr, ru, de, pl, ko, fa, ja]。

- **perPage** `Number`

  Default: `10`.

  每次加载的数据大小，最多 100。

- **distractionFreeMode** `Boolean`

  Default: false。

  类似 Facebook 评论框的全屏遮罩效果.

- **pagerDirection** `String`

  Default: 'last'

  评论排序方式， `last`为按评论创建时间倒叙，`first`为按创建时间正序。

- **createIssueManually** `Boolean`

  Default: `false`.

  如果当前页面没有相应的 isssue 且登录的用户属于 admin，则会自动创建 issue。如果设置为 `true`，则显示一个初始化页面，创建 issue 需要点击 `init` 按钮。

- **proxy** `String`

  Default: `https://cors-anywhere.azm.workers.dev/https://github.com/login/oauth/access_token`.

  GitHub oauth 请求到反向代理，为了支持 CORS。 [为什么要这样?](https://github.com/isaacs/github/issues/330)

- **flipMoveOptions** `Object`

  Default:

  ```js
    {
      staggerDelayBy: 150,
      appearAnimation: 'accordionVertical',
      enterAnimation: 'accordionVertical',
      leaveAnimation: 'accordionVertical',
    }
  ```

  评论列表的动画。 [参考](https://github.com/joshwcomeau/react-flip-move/blob/master/documentation/enter_leave_animations.md)

- **enableHotKey** `Boolean`

  Default: `true`.

  启用快捷键(cmd|ctrl + enter) 提交评论.

## 实例方法

- **render(String/HTMLElement)**

  初始化渲染并挂载插件。

## TypeScript

已经包括了配置项和 Gitalk 类的类型定义，不包括 React 组件的类型定义。

## 贡献

1. [Fork 代码仓库](https://github.com/gitalk/gitalk/fork) 并从 master 创建你的分支
2. 如果你添加的代码需要测试，请添加测试！
3. 如果你修改了 API，请更新文档。
4. 确保单元测试通过 (npm test).
5. 确保代码风格一致 (npm run lint).
6. 提交你的代码 (git commit) [提交信息格式参考](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#-git-commit-guidelines)

## 类似项目

- [gitment](https://github.com/imsun/gitment)
- [vssue](https://vssue.js.org/zh)

## 许可

MIT

[npm-version-image]: https://img.shields.io/npm/v/gitalk.svg?style=flat-square
[npm-version-url]: https://www.npmjs.com/package/gitalk
[cdnjs-version-image]: https://img.shields.io/cdnjs/v/gitalk.svg?style=flat-square
[cdnjs-version-url]: https://cdnjs.com/libraries/gitalk
[david-dm-image]: https://david-dm.org/gitalk/gitalk.svg?style=flat-square
[david-dm-url]: https://david-dm.org/gitalk/gitalk
[travis-image]: https://img.shields.io/travis/gitalk/gitalk/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/gitalk/gitalk
[coveralls-image]: https://img.shields.io/coveralls/gitalk/gitalk/master.svg?style=flat-square
[coveralls-url]: https://coveralls.io/github/gitalk/gitalk
[gzip-size]: https://img.badgesize.io/https://unpkg.com/gitalk/dist/gitalk.min.js?compression=gzip&style=flat-square
[gzip-url]: https://unpkg.com/gitalk/dist/gitalk.min.js


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.7.2](https://github.com/gitalk/gitalk/compare/v1.7.1...v1.7.2) (2021-03-03)

### Bug Fixes

- replace proxy ([68786ee](https://github.com/gitalk/gitalk/commit/68786ee5d28b1e93fc7dcb57edabc70370d80cca))

### [1.7.1](https://github.com/gitalk/gitalk/compare/v1.7.0...v1.7.1) (2021-02-28)

### Bug Fixes

- Update comment not work when post like or unlike ([#416](https://github.com/gitalk/gitalk/issues/416)) ([e0ca916](https://github.com/gitalk/gitalk/commit/e0ca91652165c0a1c54a9b214e00168e1f8dbb17))

## [1.7.0](https://github.com/gitalk/gitalk/compare/v1.6.2...v1.7.0) (2020-10-15)

### Features

- add Korean language support ([#407](https://github.com/gitalk/gitalk/issues/407)) ([8d874e0](https://github.com/gitalk/gitalk/commit/8d874e02f12282f6a28ec0958f99685cf92f5fd9))
- add Polish translations ([#406](https://github.com/gitalk/gitalk/issues/406)) ([6420f31](https://github.com/gitalk/gitalk/commit/6420f315ec92eb9b30e4532cffbb26e862237ef6))

### Bug Fixes

- replace onMouseDown to onClick ([873caa8](https://github.com/gitalk/gitalk/commit/873caa8407407476ea682bad8a5b6f60534e3cb4))

### [1.6.2](https://github.com/gitalk/gitalk/compare/v1.6.0...v1.6.2) (2020-02-29)

### Bug Fixes

- set default avatar for image load fail ([#357](https://github.com/gitalk/gitalk/issues/357)) ([49e37fd](https://github.com/gitalk/gitalk/commit/49e37fddf6b374ce863f1888d23b0e09a53b3db0))

### [1.6.1](https://github.com/gitalk/gitalk/compare/v1.6.0...v1.6.1) (2020-02-29)

## [1.6.0](https://github.com/gitalk/gitalk/compare/v1.5.2...v1.6.0) (2020-02-16)

### Features

- add german translations ([#309](https://github.com/gitalk/gitalk/issues/309)) ([f92f372](https://github.com/gitalk/gitalk/commit/f92f3722fbabce236c8e3e13202a0cb1c91de757))
- refresh after issue created. ([#277](https://github.com/gitalk/gitalk/issues/277)) ([6dfcc21](https://github.com/gitalk/gitalk/commit/6dfcc2150b1bd51f6ed939430cad7f712b33602a))

### Bug Fixes

- debounce at clicking btn to create comments ([eb35bdf](https://github.com/gitalk/gitalk/commit/eb35bdf9ad23e5f5a5240848d6465a863ffa70e0))
- repeat create comment ([b53eccb](https://github.com/gitalk/gitalk/commit/b53eccb31af582818079c944f840b82b5ac8f031))
- **comment:** comments via email notification ([d536f08](https://github.com/gitalk/gitalk/commit/d536f08437532fa5f526b7f3becec3bf97fac9dd)), closes [#224](https://github.com/gitalk/gitalk/issues/224)


src/__tests__/util.test.ts
import { queryParse, queryStringify, formatErrorMsg } from '../util'

describe('util', function () {
  const search = 'a=b&c=1'
  const searchObject = {
    a: 'b',
    c: '1',
  }

  describe('queryParse', function () {
    it(search, function () {
      expect(queryParse(search)).toEqual(searchObject)
    })
    it(`?${search}`, function () {
      expect(queryParse(`?${search}`)).toEqual(searchObject)
    })
  })

  describe('queryStringify', function () {
    it('object', function () {
      expect(queryStringify(searchObject)).toEqual(search)
    })
    it('empty value', function () {
      expect(queryStringify({ a: '' })).toEqual('a=')
    })
  })

  describe('formatErrorMsg', function () {
    it('err.response', function () {
      expect(
        formatErrorMsg({
          response: {
            data: {
              message: 'm1',
              errors: [
                {
                  message: 'm21',
                },
                {
                  message: 'm22',
                },
              ],
            },
          },
        })
      ).toEqual('Error: m1. m21, m22')
    })
    it('err.msg', function () {
      expect(
        formatErrorMsg({
          message: 'm1',
        })
      ).toEqual('Error: m1')
    })
  })
})


src/component/__tests__/svg.spec.ts
import React from 'react'
import { shallow } from 'enzyme'

import Svg from './svg'

describe('Svg', function () {
  it('set props className', function () {
    const className = 'class'
    expect(shallow(<Svg className={className} />).hasClass(`gt-ico ${className}`)).toBe(true)
  })
  it('set props text', function () {
    const text = 'text'
    expect(
      shallow(<Svg text={text} />)
        .find('.gt-ico-text')
        .text()
    ).toEqual(text)
  })
})


src/component/__tests__/action.spec.ts
import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'

import Action from './action'

describe('Action', function () {
  it('set props className', function () {
    const className = 'class'
    expect(shallow(<Action className={className} />).hasClass(className)).toBe(true)
  })
  it('set props text', function () {
    const text = 'text'
    expect(
      shallow(<Action text={text} />)
        .find('.gt-action-text')
        .text()
    ).toEqual(text)
  })
  it('set props onClick', function () {
    const onButtonClick = sinon.spy()
    const wrapper = shallow(<Action onClick={onButtonClick} />)
    wrapper.find('a').simulate('click')
    expect(onButtonClick.calledOnce).toBe(true)
  })
})


src/util.ts
import axios from 'axios'

export const queryParse = (search = window.location.search) => {
  if (!search) return {}
  const queryString = search[0] === '?' ? search.substring(1) : search
  const query: {
    [key: string]: string
  } = {}
  queryString.split('&').forEach((queryStr) => {
    const [key, value] = queryStr.split('=')
    /* istanbul ignore else */
    if (key) query[decodeURIComponent(key)] = decodeURIComponent(value)
  })

  return query
}

export const queryStringify = (query) => {
  const queryString = Object.keys(query)
    .map((key) => `${key}=${encodeURIComponent(query[key] || '')}`)
    .join('&')
  return queryString
}

export const axiosJSON = axios.create({
  headers: {
    Accept: 'application/json',
  },
})

export const axiosGithub = axios.create({
  baseURL: 'https://api.github.com',
  headers: {
    Accept: 'application/json',
  },
})

export const getMetaContent = (name: string, content = 'content') => {
  const el = window.document.querySelector(`meta[name='${name}']`)
  return el && el.getAttribute(content)
}

export const formatErrorMsg = (err) => {
  let msg = 'Error: '
  if (err.response && err.response.data && err.response.data.message) {
    msg += `${err.response.data.message}. `
    err.response.data.errors && (msg += err.response.data.errors.map((e) => e.message).join(', '))
  } else {
    msg += err.message
  }
  return msg
}

export const hasClassInParent = (element, ...className) => {
  /* istanbul ignore next */
  let yes = false
  /* istanbul ignore next */
  if (typeof element.className === 'undefined') return false
  /* istanbul ignore next */
  const classes = element.className.split(' ')
  /* istanbul ignore next */
  className.forEach((c, i) => {
    /* istanbul ignore next */
    yes = yes || classes.indexOf(c) >= 0
  })
  /* istanbul ignore next */
  if (yes) return yes
  /* istanbul ignore next */
  return element.parentNode && hasClassInParent(element.parentNode, className)
}

export function sleep(ttl: number) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(null)
    }, ttl)
  })
}


src/__mocks__/setup.js
import moxios from 'moxios'

// mock window.*
global.localStorage = {
  getItem() {
    return '{}'
  },
  setItem() {},
  removeItem() {},
}
global.document = {
  title: '',
}
global.location = {
  search: '',
  href: '',
}

// mock axios request
moxios.stubRequest(/\/repos\/.*\/comments/, {
  status: 200,
  response: [
    {
      id: 1,
      html_url: 'https://github.com/xxx/xxx/issues/1#issuecomment-xxx',
      body_html: '<p>111</p>',
      created_at: '2017-06-30T09:00:19Z',
      user: {
        login: 'booxood',
        avatar_url: 'https://avatars0.githubusercontent.com/u/2151410?v=3',
        html_url: 'https://github.com/booxood',
      },
    },
    {
      id: 2,
      html_url: 'https://github.com/xxx/xxx/issues/1#issuecomment-xxx',
      body_html: '<p>222</p>',
      created_at: '2017-06-30T09:01:19Z',
      user: {
        login: 'booxood',
        avatar_url: 'https://avatars0.githubusercontent.com/u/2151410?v=3',
        html_url: 'https://github.com/booxood',
      },
    },
  ],
})
moxios.stubRequest(/\/repos\/.*\/issues/, {
  status: 200,
  response: [
    {
      html_url: 'https://github.com/xxx/xxx/issues/1',
      comments_url: 'https://api.github.com/repos/xxx/xxx/issues/1/comments',
      comments: 2,
    },
  ],
})
moxios.stubRequest(/\/graphql/, {
  status: 200,
  response: {
    data: {
      repository: {
        issue: {
          comments: {
            nodes: [
              {
                databaseId: 1,
                author: {},
                bodyHTML: '<p>111</p>',
                body: '111',
                createdAt: '2017-06-30T09:00:19Z',
              },
              {
                databaseId: 2,
                author: {},
                bodyHTML: '<p>222</p>',
                body: '222',
                createdAt: '2017-06-30T09:01:19ZZ',
              },
            ],
            pageInfo: {},
            totalCount: 2,
          },
        },
      },
    },
  },
})
moxios.stubRequest(/.*/, {
  status: 404,
})


src/graphql/getComments.ts
import { axiosGithub } from '../util'

const getQL = (vars, pagerDirection) => {
  const cursorDirection = pagerDirection === 'last' ? 'before' : 'after'
  const ql = `
  query getIssueAndComments(
    $owner: String!,
    $repo: String!,
    $id: Int!,
    $cursor: String,
    $pageSize: Int!
  ) {
    repository(owner: $owner, name: $repo) {
      issue(number: $id) {
        title
        url
        bodyHTML
        createdAt
        comments(${pagerDirection}: $pageSize, ${cursorDirection}: $cursor) {
          totalCount
          pageInfo {
            ${pagerDirection === 'last' ? 'hasPreviousPage' : 'hasNextPage'}
            ${cursorDirection === 'before' ? 'startCursor' : 'endCursor'}
          }
          nodes {
            id
            databaseId
            author {
              avatarUrl
              login
              url
            }
            bodyHTML
            body
            createdAt
            reactions(first: 100, content: HEART) {
              totalCount
              viewerHasReacted
              pageInfo{
                hasNextPage
              }
              nodes {
                id
                databaseId
                user {
                  login
                }
              }
            }
          }
        }
      }
    }
  }
  `

  if (vars.cursor === null) delete vars.cursor

  return {
    operationName: 'getIssueAndComments',
    query: ql,
    variables: vars,
  }
}

async function getComments(issue) {
  const { owner, repo, perPage, pagerDirection, defaultAuthor } = this.options
  const { cursor, comments } = this.state
  const res = await axiosGithub
    .post(
      '/graphql',
      getQL(
        {
          owner,
          repo,
          id: issue.number,
          pageSize: perPage,
          cursor,
        },
        pagerDirection
      ),
      {
        headers: {
          Authorization: `bearer ${this.accessToken}`,
        },
      }
    )
  const data = res.data.data.repository.issue.comments
  const items = data.nodes.map((node) => {
    const author = node.author || defaultAuthor

    return {
      id: node.databaseId,
      gId: node.id,
      user: {
        avatar_url: author.avatarUrl,
        login: author.login,
        html_url: author.url,
      },
      created_at: node.createdAt,
      body_html: node.bodyHTML,
      body: node.body,
      html_url: `https://github.com/${owner}/${repo}/issues/${issue.number}#issuecomment-${node.databaseId}`,
      reactions: node.reactions,
    }
  })
  let cs
  if (pagerDirection === 'last') {
    cs = [...items, ...comments]
  } else {
    cs = [...comments, ...items]
  }
  const isLoadOver = data.pageInfo.hasPreviousPage === false || data.pageInfo.hasNextPage === false
  this.setState({
    comments: cs,
    isLoadOver,
    cursor: data.pageInfo.startCursor || data.pageInfo.endCursor,
  })
  return cs
}

export default getComments


typings/index.d.ts
// Type definitions for gitalk 1.5
// Project: https://github.com/gitalk/gitalk
// Definitions by: Chen Junda<https://github.com/viccrubs>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
import { FlipMoveProps } from 'react-flip-move'

/*~ If this module is a UMD module that exposes a global variable 'myLib' when
 *~ loaded outside a module loader environment, declare that global here.
 *~ Otherwise, delete this declaration.
 */

export as namespace Gitalk

export default Gitalk

/*~ You can declare types that are available via importing the module */
declare class Gitalk {
  /**
   * Construct a Gitalk instance.
   * @param options Gitalk options
   */
  constructor(options: Gitalk.GitalkOptions)

  /**
   * Init render and mount plugin.
   * @param element the element or the selector to the element.
   */
  render(element: string | HTMLElement): void
}

declare namespace Gitalk {
  interface GitalkOptions {
    /**
     * GitHub Application Client ID.
     */
    clientID: string

    /**
     * GitHub Application Client Secret.
     */
    clientSecret: string

    /**
     * Name of Github repository.
     */
    repo: string

    /**
     * GitHub repository owner.
     * Can be personal user or organization.
     */
    owner: string

    /**
     * GitHub repository owner and collaborators.
     * (Users who having write access to this repository)
     */
    admin: string[]

    /**
     * The unique id of the page.
     * Length must less than 50.
     *
     * @default location.href
     */
    id?: string

    /**
     * The issue ID of the page.
     * If the number attribute is not defined, issue will be located using id.
     *
     * @default -1
     */
    number?: number

    /**
     * GitHub issue labels.
     *
     * @default ['Gitalk']
     */
    labels?: string[]

    /**
     * GitHub issue title.
     *
     * @default document.title
     */
    title?: string

    /**
     * GitHub issue body.
     *
     * @default location.href + header.meta[description]
     */
    body?: string

    /**
     * Localization language key.
     * en, zh-CN and zh-TW are currently available.
     *
     * @default navigator.language || navigator.userLanguage
     */
    language?: string

    /**
     * Pagination size, with maximum 100.
     *
     * @default 10
     */
    perPage?: number

    /**
     * Facebook-like distraction free mode.
     *
     * @default false
     */
    distractionFreeMode?: boolean

    /**
     * Comment sorting direction.
     * Available values are last and first.
     *
     * @default "last"
     */
    pagerDirection?: 'last' | 'first'

    /**
     * By default, Gitalk will create a corresponding github issue for your every single page automatically when the logined user is belong to the admin users.
     * You can create it manually by setting this option to true.
     *
     * @default false
     */
    createIssueManually?: boolean

    /**
     * GitHub oauth request reverse proxy for CORS.
     * [Why need this?](https://github.com/isaacs/github/issues/330)
     *
     * @default "https://cors-anywhere.azm.workers.dev/https://github.com/login/oauth/access_token"
     */
    proxy?: string

    /**
     * Comment list animation.
     *
     * [Reference](https://github.com/joshwcomeau/react-flip-move/blob/master/documentation/enter_leave_animations.md)
     *
     * Default value:
     * ```js
     * {
     *  staggerDelayBy: 150,
     *  appearAnimation: 'accordionVertical',
     *  enterAnimation: 'accordionVertical',
     *  leaveAnimation: 'accordionVertical',
     * }
     * ```
     */
    flipMoveOptions?: FlipMoveProps

    /**
     * Enable hot key (cmd|ctrl + enter) submit comment.
     *
     * @default true
     */
    enableHotKey?: boolean
  }
}
